Onboarding Improvements: Epilogue Screen Jetpack App

ParaskP7 opened the pull request:
Parent #14990 

This PR makes sure that the existing `Login Epilogue` screen is shown for the `Jetpack` app, irrespective of the `Onboarding Improvements` feature flag. As such, below is a table based on which the new `Login Epilogue` screen is to be shown:

App | Onboarding Improvements Feature Flag | Login Epilogue Screen
-----|-----|-----
WordPress | Off | Existing
WordPress | On | New
Jetpack | Off | Existing
Jetpack | On | Existing

-----

To test:

#### Scenario: `WordPress - Onboarding Improvements Feature Flag - OFF`
1. Open the `WordPress` app.
2. Go to My Site -> App Settings -> Test feature configuration.
3. Disable the OnboardingImprovementsFeatureConfig feature flag.
4. Log-out and log-in to the `WordPress` app again.
5. Make sure the **EXISTING** `Login Epilogue` screen is shown and works as expected.

#### Scenario: `WordPress - Onboarding Improvements Feature Flag - ON`
1. Open the `WordPress` app.
2. Go to My Site -> App Settings -> Test feature configuration.
3. Enable the OnboardingImprovementsFeatureConfig feature flag.
4. Log-out and log-in to the `WordPress` app again.
5. Make sure the **NEW** `Login Epilogue` screen is shown and works as expected.

#### Scenario: `Jetpack - Onboarding Improvements Feature Flag - OFF`
1. Open the `Jetpack` app.
2. Go to My Site -> App Settings -> Test feature configuration.
3. Disable the OnboardingImprovementsFeatureConfig feature flag.
4. Log-out and log-in to the `Jetpack` app again.
5. Make sure the **EXISTING** `Login Epilogue` screen is shown and works as expected.

#### Scenario: `Jetpack - Onboarding Improvements Feature Flag - ON`
1. Open the `Jetpack` app.
2. Go to My Site -> App Settings -> Test feature configuration.
3. Enable the OnboardingImprovementsFeatureConfig feature flag.
4. Log-out and log-in to the `Jetpack` app again.
5. Make sure the **EXISTING** `Login Epilogue` screen is shown and works as expected.

-----

## Regression Notes
1. Potential unintended areas of impact

The existing `Login Epilogue` screen might be somehow affected by all the changes in this PR since all 3 screen are sharing all the logic, that is code-wise.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

See `To test` section above.

3. What automated tests I added (or what prevented me from doing so)

N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ParaskP7 committed:
Feature: Show existing epilogue screen for jetpack app

The new epilogue screen(s) is only to be showed for the WordPress app.
Since on the Jetpack app it has been decided to not allow users to
create new site (at least for now), it has been also decided that the
new epilogue screen will not be shown. This is done because the new
epilogue screen also allows for new site creation.

As such, for the Jetpack app, the quick start prompt is never shown as
well. But, since this fullscreen dialog is now shown at all, any logic
related to that and excluding it for the Jetpack app is ignored as
unreached.

ParaskP7 committed:
Refactor: Extract new login epilogue private method

This private method will help keep all the logic on whether the new
login epilogue screen should be shown or not. Later on this will get
enhance with additional logic that will incorporate the alternative flow
and show the existing login epilogue if that is the case.

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEpilogueFragment.java:
```diff
@@ -116,6 +116,10 @@ private int loginEpilogueScreenResource() {
         }
     }
 
+    private boolean newLoginEpilogue() {
```
@ParaskP7 
❓  `newLoginEpilogue` - can we come up with a more descriptive name. Our future selves will thank us.

ParaskP7 replied:
👋 @zwarm !

Thank you so much for this comment, I was actually looking forward to it! 🌟 

I spent 5' mins (ish) in thinking what would the best name be for this, but after a while, and me trying a few thing, I ended up decided on giving it a name not by the `What` it does principle, but by the `Why` it does what it does principle. My problem with that method was that it is used in a number of places and although giving a name like `shouldUseNewLoginEpilogueScreen()` might be a good fit for this one usage, it would not be a good fit for all the other usages, and vice-versa. As such, I decided to keep the naming simple using a common denominator.

I know we have discussed about changing the `New` to something else, since this is not actually a new `Epilogue Screen`, but another one and both the `Existing` and this `Another` one will be anyway both used, even when we remove the `Onboarding Improvements` feature flag, But, since everything is still named with either a `New` suffix or prefix, I chose to go with that. On a separate refactor PR we can change all that to a more appropriate naming, after we sync between us and decided what would the best naming be.

Also, I chose to stick to the `newLoginEpilogue()` naming for now since my next `Alternative Flow` PR will add to this signature and I wasn't sure how this might require a name change yet again. Thus, I want to defer this naming decision for last.

👋 @ashiagr , I am shamelessly pinging you on this as well, because:
1. Maybe you can suggest a good naming to us, and
2. So that you are aware of our already generic naming related discussion with @zwarm so that we can sync on it later on and decide on something more appropriate overall.

PS: When I was trying on different names I also tried the following:
- `isNewLoginEpilogue()`
- `withNewLoginEpilogue()`
- `whenForNewLoginEpilogue()`
- `shouldUseNewLoginEpilogueScreen()`
- ...and some others I didn't quite end up liking.
Please suggest me a better alternative and I'll be happy to rename it! 🙏

ashiagr replied:
Thanks for the ping @ParaskP7! Naming is really hard 😅. 

Based on the conditions included inside the method, we might want to rename it to  `shouldShowNewLoginEpilogueScreen()` but let's defer this decision till we've finalized all changes and are ready to remove the `new` prefix or suffix used for the feature.

ParaskP7 replied:
Naming is hard indeed! 😅 

Thanks for the suggestion @ashiagr , this `shouldShowNewLoginEpilogueScreen()` sounds great, but it still falls under this problem that I was having where although this is the correct name when used for this main usage:
```
if (shouldShowNewLoginEpilogueScreen()) {
    return R.layout.login_epilogue_screen_new;
} else {
    return R.layout.login_epilogue_screen;
}
```
This doesn't feel correct when used on some other usages, like the:
```
if (shouldShowNewLoginEpilogueScreen()) {
    mLoginEpilogueListener.onCreateNewSite();
} else {
    mLoginEpilogueListener.onContinue();
}
```
This is because on the one case is about deciding which screen to show, where the name totally makes sense, but on the other case is about deciding what action to do based on the screen, not asking if we want to `Show` the screen. Do you see my problem with this naming?

PS: No matter what, if you both feel that we can and should use the `shouldShowNewLoginEpilogueScreen()`, I am 💯 onboard. @zwarm your thoughts?

ashiagr replied:
I see. How about `isNewLoginEpilogueScreenEnabled()`?

ParaskP7 replied:
I like that much better yes! 🌟

ParaskP7 replied:
@zwarm I am going to go with what @ashiagr suggested for now and update the PR, also merge the latest `develop` changes and resolve any conflicts. Then, if we decided on a better name we can incorporate that on my next PR (expected today).

ParaskP7 replied:
This is now done here: 429b230634baf00587152d550d6bab498a62479d

zwarm replied:
👍  I'm onboard ... naming is **_hard_** . LOL.

zwarm reviewed (approved):
@ParaskP7 - All works as expected. I also tried some other Jetpack login scenarios just to triple check. 👍 

I left one question on a function name, but that won't stop my approval. If we can't come up with something more descriptive, then it stays.

ParaskP7 committed:
Refactor: Update method to is new login epilogue screen enabled

After a bit of discussion around coming up with a better naming than the
'newLoginEpilogue()', a better alternative was suggest, the
'isNewLoginEpilogueScreenEnabled'.

ParaskP7 committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into issue/14990-jetpack-app-epilogue-screen

ParaskP7 committed:
Fix: Remove extra is new login epilogue screen enabled for tracking

This is actually a fix since the tracking should not happen if the
'mLoginEpilogueListener' is null, which actually should never be the
case and might result to a bug. As such, when such case arise (if ever)
and given the fact that the main action(s) will not be performed
('onCreateNewSite()' or 'onContinue()'), when that happens the tracking
shouldn't be performed as well.

PS: The previous version of 'setupBottomButton(...)', the one before
'Onboarding Improvements' were introduced to the 'Login Epilogue' screen
was also having the very same problem.

ParaskP7 commented:
@zwarm and @ashiagr I also fixed an issue with the tracking logic on `setupBottomButton(...)` after updating with `develop` since I needed to incorporate this change as well.

Ashita, I have already discussed that with Annmarie yesterday and she is good to go with it. If you can double check as well and give a 👍  I can then merge this to `develop` to continue with the `Alternative Flow` PR. Many thanks for your help here! 🙏 

PS: The commit is the following: c48d5f88f5f2a7856be71d82f85946d88f698dc9

ashiagr commented:
👍 from my side. Feel free to merge it when CI completes.
Good job.

ParaskP7 commented:
> 👍 from my side. Feel free to merge it when CI completes.

Thank you! 🥇 

> Good job.

Back to you! ❤️ 

PS: As always CI complains falsely on Lint! 🤷 I'll keep re-triggering until it gets unstuck... 😅
